### PR TITLE
[SYCL] Add invalid kernel handling

### PR DESCRIPTION
**Pull request by KrystianChmielewski**
This commit adds support for handling invalid kernels.

Sometimes kernel cannot be generated for specific platform, when
trying to append launch of such faulty kernel exception is now
thrown - informing why used kernel is invalid.

Signed-off-by: Krystian Chmielewski <krystian.chmielewski@intel.com>

### CODE REVIEW

**Comment by romanovvlad** [comment]
BTW. It would be nice to have a test.

**Comment by smaslov-intel** [comment]
> This commit adds support for handling invalid kernels.
> 
> Sometimes kernel cannot be generated for specific platform, when trying to append launch of such faulty kernel exception is now thrown - informing why used kernel is invalid.
> 
> Signed-off-by: Krystian Chmielewski [krystian.chmielewski@intel.com](mailto:krystian.chmielewski@intel.com)

What is the motivating issue? Are you going to add an E2E test too?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -5418,12 +5421,28 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
+	ZE_CALL(zeKernelGetSourceAttributes, (Kernel->ZeKernel, &sourceAttributesLength, nullptr));
+	char *sourceAttributes = new char[sourceAttributesLength];
+	ZE_CALL(zeKernelGetSourceAttributes, (Kernel->ZeKernel, &sourceAttributesLength, &sourceAttributes));
+	const std::string attributesString(sourceAttributes, sourceAttributes + sourceAttributesLength);
+	delete[] sourceAttributes;
+
+	const std::regex invalidKernelAttributeRegex("invalid_kernel\\(.*\\)");
+	std::smatch invalidKernelAttributeMatch;
+	if(std::regex_search(attributesString, invalidKernelAttributeMatch, invalidKernelAttributeRegex)){
+		throw std::invalid_argument(invalidKernelAttributeMatch[0].str());
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5441]: The plugin is not supposed to throw exceptions, instead an error code should be returned to SYCL RT. The easiest is to use PI_ERROR_PLUGIN_SPECIFIC_ERROR, see existing uses

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -5418,12 +5421,28 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
     // submitted only when the comamndlist is closed. Then, a lock is held.
-    ZE_CALL(zeCommandListAppendLaunchKernel,
+    zeResult = ZE_CALL_NOCHECK(zeCommandListAppendLaunchKernel,
             (CommandList->first, Kernel->ZeKernel, &ZeThreadGroupDimensions,
              ZeEvent, (*Event)->WaitList.Length,
              (*Event)->WaitList.ZeEventList));
   }
 
+  if(zeResult != ZE_RESULT_SUCCESS){
+	uint32_t sourceAttributesLength = 0U;
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5431]: Please follow code conventions:
```suggestion
	uint32_t SourceAttributesLength = 0U;
```

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -5418,12 +5421,28 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
     // because the kernels are not actually submitted yet. Kernels will be
     // submitted only when the comamndlist is closed. Then, a lock is held.
-    ZE_CALL(zeCommandListAppendLaunchKernel,
+    zeResult = ZE_CALL_NOCHECK(zeCommandListAppendLaunchKernel,
             (CommandList->first, Kernel->ZeKernel, &ZeThreadGroupDimensions,
              ZeEvent, (*Event)->WaitList.Length,
              (*Event)->WaitList.ZeEventList));
   }
 
+  if(zeResult != ZE_RESULT_SUCCESS){
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5430]: should you only do the search below for specific L0 error code?

**Comment by KrystianChmielewski** [comment]
@smaslov-intel  "What is the motivating issue? Are you going to add an E2E test too?" 

Kernel's compiled from spirv using fp64 on platforms without fp64 support are poisoned (code is replaced) with printf. This is done to enable building module with fp64 kernels on platforms without fp64 support. This change aims to catch this and return appropriate error to user, when encountering poisoned kernel.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (KrystianChmielewski) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5430]: sure

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (KrystianChmielewski) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5431]: ok

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (KrystianChmielewski) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5441]: sure

**Comment by smaslov-intel** [comment]
> Kernel's compiled from spirv using fp64 on platforms without fp64 support are poisoned (code is replaced) with printf.  This is done to enable building module with fp64 kernels on platforms without fp64 support.

Why does a platform need the support fp64 to just **build** kernel, not to run it?

> This change aims to catch this and return appropriate error to user, when encountering poisoned kernel.

Are you saying that the regexp you are trying to match is something that is printed with that printf? If so, it looks like a hack to me. Isn't there a more robust solution to the issue?

**Comment by smaslov-intel** [comment]
> That's how intel-graphics-compiler works. The problem is if user tries to run this kernel. It would be much better to get an error instead of executing said kernel and getting information from result of kernel's printf.

This is for AOT compilation, right? Wouldn't it be much more friendly to give an error at AOT compilation-time (and request JIT compilation) rather than deferring to run-time? Mis-compiling the kernel (which printf is essentially doing) and then trying to understand/report this at run-time looks weird at least.

> Compiler appends attribute to the kernel "invalid_kernel(reason)". Regexp that I am matching is one of kernel's attributes

**Comment by KrystianChmielewski** [comment]
>Wouldn't it be much more friendly to give an error at AOT compilation-time (and request JIT compilation) rather than deferring to run-time?
This is a must to allow same build DL frameworks for PVC/ATS-M. 
The requirements for it to work are:
1. Build time: no failures should be reported. - do not give error during AOT compilation
2. Run time: execution on pvc/ats-m should report error in case application uses fp64. In AOT case 'assert' should be injected. This is done by passing "invalid_kernel" attribute.

>Is it possible for user to insert "attribute" in source code and then have a false positive diagnostic in case of a coincidental regexp match?
It's impossible to have false positive. Regexp only matches attribute "invalid_kernel(reason)"

**Comment by smaslov-intel** [comment]
> > Wouldn't it be much more friendly to give an error at AOT compilation-time (and request JIT compilation) rather than deferring to run-time?
> 
> This is a must to allow same build DL frameworks for PVC/ATS-M. The requirements for it to work are:
> 
> 1. Build time: no failures should be reported. - do not give error during AOT compilation
> 2. Run time: execution on pvc/ats-m should report error in case application uses fp64. In AOT case 'assert' should be injected. This is done by passing "invalid_kernel" attribute.

What is the benefit of pretending we have the right kernel at build-time (where we already know it is not going to work) and report an error at run-time?

> > Is it possible for user to insert "attribute" in source code and then have a false positive diagnostic in case of a coincidental regexp match?
> 
> It's impossible to have false positive. Regexp only matches attribute "invalid_kernel(reason)"

What if something else (user) injects the matching attribute?

**Comment by KrystianChmielewski** [comment]
> What is the benefit of pretending we have the right kernel at build-time (where we already know it is not going to work) and report an error at run-time?

We can use single spirv with multiple kernels (including those using fp64) and compile it to native binary for every device.

> What if something else (user) injects the matching attribute?

If user injects "invalid_kernel(reason)" attribute then PI_ERROR_PLUGIN_SPECIFIC_ERROR error with message description "invalid_kernel(reason)" will be returned as expected.

**Comment by smaslov-intel** [comment]
> We can use single spirv with multiple kernels (including those using fp64) and compile it to native binary for every device.

But you said just earlier that you can't compile fp64 code correctly if host devices doesn't support fp64. So, what is the purpose of not giving an error in compile time if the fp64 kernel is miscompiled? Maybe, you have a design doc explaining the approach in details?

> If user injects "invalid_kernel(reason)" attribute then PI_ERROR_PLUGIN_SPECIFIC_ERROR error with message description "invalid_kernel(reason)" will be returned as expected.

Why is this "as expected"? I think users can expect completely something different, and may not be happy with RT reporting error in this case. There is no compiler reserved attributes that app can't use, are there?

**Comment by KrystianChmielewski** [comment]
>  Maybe, you have a design doc explaining the approach in details?

Please look into jira XDEPS-4268.

> Why is this "as expected"? I think users can expect completely something different, and may not be happy with RT reporting error in this case. There is no compiler reserved attributes that app can't use, are there?

If user adds attribute "reqd_work_group_size(X,Y,Z)" then it's expected for work group size to be (X, Y, Z) (any other work group size on setGroupSize call will result in an error). If user adds attribute "invalid_kernel(reason)" then it's expected to return an error on appendLaunchKernel

**Comment by smaslov-intel** [comment]
> If user adds attribute "reqd_work_group_size(X,Y,Z)" then it's expected for work group size to be (X, Y, Z) (any other work group size on setGroupSize call will result in an error). If user adds attribute "invalid_kernel(reason)" then it's expected to return an error on appendLaunchKernel

Where is this behavior documented so that users know that?

**Comment by KrystianChmielewski** [comment]
@smaslov-intel where should it be documented?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -5445,19 +5447,43 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
...
-            (CommandList->first, Kernel->ZeKernel, &ZeThreadGroupDimensions,
-             ZeEvent, (*Event)->WaitList.Length,
-             (*Event)->WaitList.ZeEventList));
+    ZeResult = ZE_CALL_NOCHECK(
+        zeCommandListAppendLaunchKernel,
+        (CommandList->first, Kernel->ZeKernel, &ZeThreadGroupDimensions,
+         ZeEvent, (*Event)->WaitList.Length, (*Event)->WaitList.ZeEventList));
+  }
+
+  if (ZeResult != ZE_RESULT_SUCCESS) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5465]: Please add a TODO comment that we need to re-do this in a robust way.

**Review by smaslov-intel — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (bader) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5465]: We decided to abandon this patch and rely on the [optional kernel features](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features) implementation.
Sorry for the mess with closing/opening this PR.

**Pull request closed without merging**